File: naturals/naturals.ts
interface Nat {
    zero?: "0";
    suc?: Nat;
}

const zero : Nat = {zero: "0"};
const one : Nat = {suc : zero}
const two : Nat = {suc : one}
console.log(zero);
console.log(one);
console.log(two);


type  ONat = {
    zero : "0";
    suc?: never;
} | {
    zero?: never; 
    suc: ONat;
}

const ozero : ONat = {zero: "0"};
const oone : ONat = {suc : ozero}
const otwo : ONat = {suc : oone}
console.log(ozero);
console.log(oone);
console.log(otwo);

function plusOne(num: ONat) : ONat{
    return {suc : num};
}

function add(num1: ONat, num2: ONat) : ONat{
    if(num1.zero !== undefined) return num2;
    //if(num2.zero !== undefined) return num1;
    return {suc : add(num1.suc, num2)};
}

function leq(num1: ONat, num2: ONat) : boolean {
    if(num1.zero !== undefined) return true;
    if(num2.zero !== undefined) return false;
    return leq(num1.suc, num2.suc);
}

function eq(num1: ONat, num2: ONat) : boolean {
    if(num1.zero !== undefined) {
        if(num2.zero !== undefined) return true;
        return false;
    }
    if(num2.zero !== undefined) return false;
    return eq(num1.suc, num2.suc);
}

function addCommut(num1: ONat, num2: ONat) : boolean {
    if(num1.zero !== undefined) return (eq(add(num1, num2), add(num2, num1)));
    if(num2.zero !== undefined) return (eq(add(num1, num2), add(num2, num1)));
    else return addCommut(num1.suc, plusOne(num2));
}

console.log(plusOne(ozero));
console.log(add(otwo, oone));
console.log(leq(otwo, oone));
console.log(leq(oone, otwo));
console.log(eq(otwo, oone));
console.log(eq(otwo, otwo));
console.log(addCommut(oone, ozero));

File: naturals/naturals.rs
// Does not work due to referencing local values in functions
/*use std::fmt;

#[derive(Eq, Clone, Copy)]
enum Nat<'a> {
    Zero,
    Suc(&'a Nat<'a>)
}

impl PartialEq for Nat<'_> {
    fn eq<'a>(&self, other: &'a Nat<'_>) -> bool {
        match (*self, other) {
            (Nat::Zero, Nat::Zero) => true,
            (Nat::Zero, Nat::Suc(_)) => false,
            (Nat::Suc(_), Nat::Zero) => false,
            (Nat::Suc(ref l), Nat::Suc(ref r)) => l == r,
        }
    }
}

impl fmt::Display for Nat<'_> {
    fn fmt<'a>(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Nat::Zero => write!(f, "zero"),
            Nat::Suc(ref p) => write!(f, "suc {}", p),
        }
    }
}

fn addOne<'a>(n: &'a Nat<'_>) -> Nat<'a>{
    return Nat::Suc(&n);
}

fn add<'a>(n: &'a Nat<'_>, m: &'a Nat<'_>) -> Nat<'a>{
    match n {
        Nat::Zero => *m,
        Nat::Suc(ref p) => add(p, &addOne(&m)),
    }
}

fn main() {
    let zero = Nat::Zero;
    let one = Nat::Suc(&zero);
    println!("{}", zero);
    println!("{}", one);
    println!("{}", addOne(&zero));
    println!("{}", one == addOne(&zero));
}*/

use std::fmt;
use std::boxed::Box;

#[derive(Eq, Clone)]
enum Nat {
    Zero,
    Suc(Box<Nat>)
}

impl fmt::Display for Nat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Nat::Zero => write!(f, "zero"),
            Nat::Suc(ref p) => write!(f, "suc {}", p),
        }
    }
}

impl PartialEq for Nat {
    fn eq(&self, other: &Nat) -> bool {
        match (self, other) {
            (Nat::Zero, Nat::Zero) => true,
            (Nat::Zero, Nat::Suc(_)) => false,
            (Nat::Suc(_), Nat::Zero) => false,
            (Nat::Suc(ref l), Nat::Suc(ref r)) => l == r,
        }
    }
}

fn leq(n: &Nat, m: &Nat) -> bool {
    match (n, m) {
        (Nat::Zero, _) => true,
        (Nat::Suc(_), Nat::Zero) => false,
        (Nat::Suc(ref l), Nat::Suc(ref r)) => leq(l, r),
    }
}

fn addOne(n: &Nat) -> Nat{
    return Nat::Suc(Box::new(n.clone()));
}

fn add(n: &Nat, m: &Nat) -> Nat {
    match n {
        Nat::Zero => m.clone(),
        Nat::Suc(ref p) => Nat::Suc(Box::new(add(p, m))),
    }
}

fn mult(n: &Nat, m: &Nat) -> Nat {
    match n {
        Nat::Zero => Nat::Zero,
        //Nat::Suc(Box<Nat::Zero>) => m.clone(),
        Nat::Suc(ref p) => add(&m, &mult(p, m)),
    }
}

fn main() {
    let zero = Nat::Zero;
    let one = Nat::Suc(Box::new(zero.clone()));
    let two = Nat::Suc(Box::new(one.clone()));
    
    println!("Zero: {}", zero);
    println!("One: {}", one);
    println!("Two: {}", two);
    
    println!("0+1 = {}", addOne(&zero));
    println!("1 == 0+1? {}", one == addOne(&zero));
    println!("2 == 1+1? {}", two == add(&one, &one));
    println!("2 >= 1? {}", leq(&two, &one));

    println!("2*2 = {}", mult(&two, &two));
    println!("1 <= 2*2? {}", leq(&two, &one));
}
